python/src/robyn/calibration/media_effect_calibration.py

```diff
@@ -0,0 +1,206 @@
+from typing import List
+import pandas as pd
+import numpy as np
+from robyn.data.entities.mmmdata import MMMData
+from robyn.data.entities.hyperparameters import Hyperparameters
+from robyn.data.entities.calibration_input import CalibrationInput, ChannelCalibrationData
+from robyn.data.entities.enums import AdstockType, CalibrationScope
+from robyn.calibration.media_transformation import MediaTransformation
+from dataclasses import dataclass
+from typing import Dict, List
+import numpy as np
+
+
+@dataclass(frozen=True)
+class CalibrationResult:
+    """
+    Immutable data class that holds the results of a calibration run.
+
+    Attributes:
+        channel_scores (Dict[str, float]): MAPE scores per channel
+        calibration_constraint (float): Acceptable error range (0.01-0.1)
+        calibrated_models (List[str]): Model IDs that passed calibration
+    """
+
+    channel_scores: Dict[str, float]
+    calibration_constraint: float
+    calibrated_models: List[str]
+
+    def get_mean_mape(self) -> float:
+        """Returns mean absolute percentage error across all channels."""
+        return np.mean(list(self.channel_scores.values()))
+
+    def is_model_calibrated(self, mape_threshold: float = 0.1) -> bool:
+        """Checks if model meets calibration criteria."""
+        return self.get_mean_mape() <= mape_threshold
+
+    def __str__(self) -> str:
+        scores_str = "\n".join(f"  {channel}: MAPE = {score:.2%}" for channel, score in self.channel_scores.items())
+        return (
+            f"CalibrationResult(\n"
+            f"  Mean MAPE: {self.get_mean_mape():.2%}\n"
+            f"  Constraint: {self.calibration_constraint}\n"
+            f"  Channel Scores:\n{scores_str}\n"
+            f"  Calibrated: {self.is_model_calibrated()}\n"
+            f")"
+        )
+
+
+class MediaEffectCalibrator:
+    """
+    Handles the calibration process for media mix models by comparing model predictions
+    against actual experimental results.
+    """
+
+    def __init__(self, mmm_data: MMMData, hyperparameters: Hyperparameters, calibration_input: CalibrationInput):
+        """Initialize calibration with model data and parameters."""
+        self.mmm_data = mmm_data
+        self.hyperparameters = hyperparameters
+        self.calibration_input = calibration_input
+        self.media_transformation = MediaTransformation(hyperparameters)
+
+        # Convert date column to datetime if it's not already
+        date_col = self.mmm_data.mmmdata_spec.date_var
+        if not pd.api.types.is_datetime64_any_dtype(self.mmm_data.data[date_col]):
+            self.mmm_data.data[date_col] = pd.to_datetime(self.mmm_data.data[date_col])
+
+        self._validate_inputs()
+
+    def _split_channel(self, channel: str) -> List[str]:
```

Comment:
calibration_input should contain individual channels. See if we can support combination of channel in calibration input itself.

For example - currently calibration_input is Dict[channel:, channelCalibrationData]. If we can have Dict[List[channel], channelCalibrationData].

Reply:
something like this?

calibration_data = {
    ["facebook"]: ChannelCalibrationData(  # Single channel as a list
        lift_start_date=pd.Timestamp("2023-01-01"),
        lift_end_date=pd.Timestamp("2023-01-31"),
        lift_abs=1000,
        spend=5000
    ),
    ["search", "tv"]: ChannelCalibrationData(  # Multiple channels as a list
        lift_start_date=pd.Timestamp("2023-02-01"),
        lift_end_date=pd.Timestamp("2023-02-28"),
        lift_abs=2000,
        spend=8000
    )
}
calibration_input = CalibrationInput(channel_data=calibration_data)

Reply:
Exactly.